v2/pkg/engine/datasource/graphql_datasource/graphql_subscription_client.go

```diff
@@ -496,35 +498,43 @@ func (c *subscriptionClient) newWSConnectionHandler(requestContext, engineContex
 		return nil, fmt.Errorf("failed to dial connection")
 	}
 
-	connectionInitMessage, err := c.getConnectionInitMessage(requestContext, options.URL, options.Header)
+	// conn is not nil. Any errored return below could lead to a leaking connection.
+	// To avoid this, close connection if failure happened.
+	defer func() {
+		if err != nil {
+			conn.Close()
+		}
+	}()
+
+	initMsg, err := c.getConnectionInitMessage(requestContext, options.URL, options.Header)
 	if err != nil {
 		return nil, err
 	}
 
 	if len(options.InitialPayload) > 0 {
-		connectionInitMessage, err = jsonparser.Set(connectionInitMessage, options.InitialPayload, "payload")
```

Comment:
Does this need to be renamed? It's fine, but it seems unnecessary

Reply:
There is a global variable named `connectionInitMessage` that was shadowed in this function.